source/slang/ir-union.cpp

```diff
@@ -0,0 +1,776 @@
+        // layout.
+        //
+        TaggedUnionTypeLayout* taggedUnionTypeLayout;
+
+        // The basic approach we will use 16-byte chunks (represented as an array
+        // of `uint4`s) to reprent the "bulk" of a type, and then use a single field
+        // that could be up to 12 bytes to represent the "rest" of the type.
+        //
+        // Note that there are deeply ingrained assumptions here that all types
+        // are at least four bytes in size (so that unions cannot easily
+        // accomodate `half` value), and that any types *larger* than four bytes
+        // will need to be loaded/stored via multiple 4-byte loads/stores.
+        //
+        // With the basic idea out of the way, we need an IR level field
+        // in our struct to hold the bulk data, which comprises a "key" for
+        // looking up the field, and the type of the field itself. We also
+        // keep track of how many bytes we put in our bulk storage.
+        //
+        // The bulk field might be:
+        //
+        // - null, if none of the case types was 16 bytes or more
+        // - a single `uint4` for between 16 and 31 (inclusive) bytes
+        // - an array of `uint4`s for 32 or more bytes
+        //
+        UInt64  bulkSize = 0;
+        IRInst* bulkFieldKey = nullptr;
+        IRType* bulkFieldType = nullptr;
+
+        // The same basic idea then applies to the rest of the data.
+        //
+        // The "rest" field will be either be absent (if the size of the
+        // type was evently divisible by 16), a scalar `uint`, or else
+        // a 2- or 3-component vector of `uint`.
+        //
+        UInt64  restSize = 0;
+        IRInst* restFieldKey = nullptr;
+        IRType* restFieldType = nullptr;
+
+        // Finally, since we are currently working with tagged unions,
+        // we need a field to hold the tag, which will always be allocated
+        // after the fields that hold the bulk/rest of the payload.
+        //
+        // This field is always a single `uint`.
+        //
+        // TODO: if/when we support untagged unions, they could be handled
+        // by having this field be null.
+        //
+        IRInst* tagFieldKey;
+    };
+
+    // We will build up a list of all the tagged union types we encounter,
+    // so that we can replace them with the synthesized types when we are done.
+    //
+    List<RefPtr<TaggedUnionInfo>> taggedUnionInfos;
+
+    // It is possible that we will see the same tagged union type referenced
+    // many times in the IR, but we only want to synthesize the information
+    // above (including the various IR structures) once, so we also maintain
+    // a map from the original IR type to the corresponding information.
+    //
+    Dictionary<IRInst*, TaggedUnionInfo*> mapIRTypeToTaggedUnionInfo;
+
+    // We will process all instructions in the module in a single recursive walk.
+    //
+    void processInstRec(IRInst* inst)
```

Comment:
Yes, it could have been written that way instead. I'm not sure the recursive find would be re-usable since the set of instructions you want to look for would be different across passes. I'd consider this a "nice to have" sort of cleanup, but not critical.